servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```
@@ -1478,8 +1517,182 @@ public void setState(long fileId, SetStateOptions options)
+      }
+      mWorkerToAsyncPersistFiles.get(workerId).add(fileId);
+    }
+
+    // TODO(yupeng) TACHYON-1456 add fault tolerance and flush journal
+
+    return workerId;
+  }
+
+  /**
+   * Gets a worker where the given file is stored.
+   *
+   * @param fileId the file id, -1 if no worker can be found
+   * @return the storing worker
+   * @throws FileDoesNotExistException when the file does not exist on any worker
+   */
+  private long getWorkerStoringFile(long fileId) throws FileDoesNotExistException {
+    Map<Long, Integer> workerBlockCounts = Maps.newHashMap();
+    List<FileBlockInfo> blockInfoList;
+    try {
+      blockInfoList = getFileBlockInfoList(fileId);
+
+      for (FileBlockInfo fileBlockInfo : blockInfoList) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          if (workerBlockCounts.containsKey(blockLocation.workerId)) {
+            workerBlockCounts.put(blockLocation.workerId,
+                workerBlockCounts.get(blockLocation.workerId) + 1);
+          } else {
+            workerBlockCounts.put(blockLocation.workerId, 1);
+          }
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    } catch (InvalidPathException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    if (workerBlockCounts.size() == 0) {
+      LOG.error("The file " + fileId + " does not exist on any worker");
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    // TODO(yupeng) remove the requirement that all the blocks of a file must be stored on the same
+    // worker
+    // return the first worker that has all the blocks
+    for (Entry<Long, Integer> entry : workerBlockCounts.entrySet()) {
+      if (entry.getValue() == blockInfoList.size()) {
+        return entry.getKey();
+      }
+    }
```

The previous loop could be written in a fail fast way (and this loop would not be necessary), is it done this way to make it easier to modify in the future?